### PR TITLE
Add SwaggerUi to replace Apiary

### PR DESCRIPTION
**Pull request by blacksam07**
#### Board:
* [Ticket Replace Apiary with self hosted SwaggerUI](https://www.notion.so/rootstrap/Replace-Apiary-with-self-hosted-SwaggerUI-ba10e2bd0aa9418486f1b34f34b4b3a4)
---
#### Description:
- Use Rswag to expose generated swagger documentation
- Expose `api-docs` endpoint with `basic-auth`

---
#### Notes:
* Create a [PR](https://github.com/exoego/rspec-openapi/pull/147) in `rspec-openapi` to include the project title 
---
#### Tasks:
- [X] Expose generate `openapi.yaml` doc with SwaggerUI
- [X] The API Docs can be accessed through the Rails app
- [X] The API Docs are not publicly accessible

---
#### Risk:
*
---
#### Preview:
image

### CODE REVIEW

File: config/initializers/rswag_ui.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+Rswag::Ui.configure do |c|
+  # List the Swagger endpoints that you want to be documented through the
+  # swagger-ui. The first parameter is the path (absolute or relative to the UI
+  # host) to the corresponding endpoint and the second is a title that will be
+  # displayed in the document selector.
+  # NOTE: If you're using rspec-api to expose Swagger files
+  # (under swagger_root) as JSON or YAML endpoints, then the list below should
+  # correspond to the relative paths for those endpoints.
+
+  c.swagger_endpoint "/api-docs/openapi#{ENV.fetch('TEST_ENV_NUMBER', '')}.yaml", 'API V1 Docs'
```
Comment (santib) [inline config/initializers/rswag_ui.rb:12]:
```suggestion
  c.swagger_endpoint "/api-docs/openapi.yaml", 'API V1 Docs'
```

I don't think we need to care about test parallelization here

**Comment by santib** [comment]
image

this link is not working properly

File: doc/openapi.yaml [hunk 2]
```diff
@@ -19,7 +19,7 @@ paths:
         example: '1.0'
       responses:
         '200':
-          description: returns no need to update
```
Comment (santib) [inline doc/openapi.yaml:22]: you don't have to commit this file, it gets automatically regenerated

File: package-lock.json [hunk 3]
```diff
@@ -0,0 +1,632 @@
+{
+  "name": "rails_api_base",
```
Comment (santib) [inline package-lock.json:2]: you don't have to commit this file, we use `yarn` not `npm`

File: yarn.lock [hunk 4]
```diff
@@ -3,138 +3,33 @@
 
 
 "@activeadmin/activeadmin@3.1.0":
-  version "3.1.0"
-  resolved "https://registry.yarnpkg.com/@activeadmin/activeadmin/-/activeadmin-3.1.0.tgz#db3d6179ff55668d667c0139f85ed09399d8f6e3"
-  integrity sha512-BWKQGs5H5VHvo47Wje5aNEnir85/TnKL2jqshCsXUhwoFCm3OyqM7Tu2yeS0E2h32QAwbCO9HPLf5hIzVSYJ2A==
+  "integrity" "sha512-BWKQGs5H5VHvo47Wje5aNEnir85/TnKL2jqshCsXUhwoFCm3OyqM7Tu2yeS0E2h32QAwbCO9HPLf5hIzVSYJ2A=="
```
Comment (santib) [inline yarn.lock:6]: I think these changes are unrelated and shouldn't be committed

**Comment by santib** [comment]
Apiary should no longer be mentioned in the README. Also, please add something in `docs/api_documentation.md` about the API Docs UI

File: config/initializers/rswag_ui.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+  # Add Basic Auth in case your API is private
+  c.basic_auth_enabled = true
```
Comment (JulianPasquale) [inline config/initializers/rswag_ui.rb:15]: What if we have both? public and private endpoints?

File: config/initializers/rswag_ui.rb (hunk 1 above)
Comment (blacksam07) [reply config/initializers/rswag_ui.rb:12]: @santib sure, I was use it because the file are defined with this value

File: doc/openapi.yaml (hunk 2 above)
Comment (blacksam07) [reply doc/openapi.yaml:22]: I included this file to update it

File: package-lock.json (hunk 3 above)
Comment (blacksam07) [reply package-lock.json:2]: Ok no problem

File: config/initializers/rswag_ui.rb (hunk 5 above)
Comment (blacksam07) [reply config/initializers/rswag_ui.rb:15]: Is the way that `rswag` works, we need to define the API endpoints that will use it for UI section

File: config/initializers/rswag_api.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+Rswag::Api.configure do |c|
+  # Specify a root folder where Swagger JSON files are located
+  # This is used by the Swagger middleware to serve requests for API descriptions
+  # NOTE: If you're using rswag-specs to generate Swagger, you'll need to ensure
+  # that it's configured to generate files in the same folder
+  c.swagger_root = Rails.root.to_s + '/doc'
```
Comment (santib) [inline config/initializers/rswag_api.rb:8]:
```suggestion
  c.swagger_root = "#{Rails.root}/doc"
```

I think this is the same

File: config/routes.rb [hunk 7]
```diff
@@ -1,6 +1,8 @@
 # frozen_string_literal: true
 
 Rails.application.routes.draw do
+  mount Rswag::Ui::Engine => '/api-docs'
+  mount Rswag::Api::Engine => '/api-docs'
```
Comment (santib) [inline config/routes.rb:5]: minor comment: maybe move this to the bottom of the file? Rails reads routes in order, I prefer to keep these endpoint at the bottom

File: doc/openapi.yaml (hunk 2 above)
Comment (santib) [reply doc/openapi.yaml]: this file is autogenerated in the `main` branch, check https://github.com/rootstrap/rails_api_base/blob/main/.github/workflows/ci.yml#L188-L195

Comment (blacksam07) [reply doc/openapi.yaml]: Yeah, was my mistake

File: config/routes.rb (hunk 7 above)
Comment (blacksam07) [reply config/routes.rb:5]: No problem, I prefer to set mount at first of the file but I can move it at end

**Review by santib — APPROVED** [review]
Great job, thanks for bringing this improvement to the `rails_api_base`.

Last thing: remove the reference to Apiary in the `README.md`

Thanks!

File: config/initializers/rswag_ui.rb (hunk 5 above)
Comment (JulianPasquale) [reply config/initializers/rswag_ui.rb:15]: Not sure if that answers my question.
If we build an app that has public endpoints, we need to also have a "public" documentation that anyone can see in order to understand how the API works.

So the question is, is it possible to only scope the auth to a subset of our endpoints and not the whole app?

File: yarn.lock [hunk 8]
```diff
@@ -174,4 +174,4 @@ jquery-ujs@1.2.3, jquery-ujs@^1.2.2:
 jquery@3.7.1, "jquery@>=1.8.0 <4.0.0", jquery@^3.4.1:
   version "3.7.1"
   resolved "https://registry.yarnpkg.com/jquery/-/jquery-3.7.1.tgz#083ef98927c9a6a74d05a6af02806566d16274de"
-  integrity sha512-m4avr8yL8kmFN8psrbFFFmB/If14iN5o9nw/NgnnM+kybDJpRsAynV2BsfpTYrTRysYUdADVD7CkUUizgkpLfg==
+  integrity sha512-m4avr8yL8kmFN8psrbFFFmB/If14iN5o9nw/NgnnM+kybDJpRsAynV2BsfpTYrTRysYUdADVD7CkUUizgkpLfg==
```
Comment (szcrubyrootstrap) [inline yarn.lock:177]: missing white space

File: README.md [hunk 9]
```diff
@@ -117,9 +118,8 @@ To illustrate, `bin/rails console` will run the console in the docker container
 
 ## Api Docs
 
-https://railsapibasers.docs.apiary.io/
-
-With [RSpec API Doc Generator](https://github.com/exoego/rspec-openapi) you can generate the docs after writing requests specs.
+- [RSpec API Doc Generator](https://github.com/exoego/rspec-openapi) you can generate the docs after writing requests specs.
+- [Rswag](https://github.com/rswag/rswag) you can expose the generate docs.
```
Comment (guillermoap) [inline README.md:122]:
```suggestion
- [Rswag](https://github.com/rswag/rswag) you can expose the generated docs.
```

**Review by guillermoap — APPROVED** [review]

File: config/initializers/rswag_ui.rb (hunk 5 above)
Comment (blacksam07) [reply config/initializers/rswag_ui.rb:15]: @JulianPasquale Oh Oh, I understood that you are asking about `rswag` sorry

- I'm not sure, but I looked in the Swagger documentation and didn't find anything about adding auth only for a group of endpoints, also following the ticket scope I included the authorization here, but we can disable it by setting the value to "false". `

Comment (JulianPasquale) [reply config/initializers/rswag_ui.rb:15]: Cool. I think it would be nice to explore it so we can allow the mix of public and private docs. Not a blocker since auth can be disabled 👍

**Review by JulianPasquale — APPROVED** [review]

File: config/initializers/rswag_ui.rb (hunk 5 above)
Comment (santib) [reply config/initializers/rswag_ui.rb:15]: I think this is a good discussion, maybe we can follow up in a meeting? I think what we need to answer is: who are we expecting to be the consumers of these endpoints?
- Frontend devs?
- Backend devs from other teams/apps?
- External products/services?

**Review by juanmferreira93 — APPROVED** [review]

**Review by szcrubyrootstrap — APPROVED** [review]
